pkg/pricer/pricer.go

```diff
@@ -218,9 +309,45 @@ func (s *Pricer) DefaultPriceTable() []uint64 {
 }
 
 func (s *Pricer) DefaultPrice(PO uint8) uint64 {
-	neighborhoodDepth := s.topology.NeighborhoodDepth()
+	neighborhoodDepth := s.neighborhoodDepth()
 	if PO > neighborhoodDepth {
 		PO = neighborhoodDepth
 	}
 	return uint64(neighborhoodDepth-PO+1) * s.poPrice
 }
+
+func (s *Pricer) neighborhoodDepth() uint8 {
+	var neighborhoodDepth uint8
+	if s.topology != nil {
+		neighborhoodDepth = s.topology.NeighborhoodDepth()
+	}
+	return neighborhoodDepth
+}
+
+func (s *Pricer) PriceHeadler(receivedHeaders p2p.Headers, peerAddress swarm.Address) (returnHeaders p2p.Headers) {
+
+	chunkAddress, receivedPrice, err := headerutils.ReadPricingHeaders(receivedHeaders)
+	if err != nil {
+		return p2p.Headers{
+			"error": []byte("Error reading pricing headers"),
+		}
+	}
+
+	s.logger.Debugf("price headler: received target %v with price as %v, from peer %s", chunkAddress, receivedPrice, peerAddress)
+	checkPrice, index := s.PriceWithIndexForPeer(peerAddress, chunkAddress)
+
+	returnHeaders, err = headerutils.MakePricingResponseHeaders(checkPrice, chunkAddress, index)
+	if err != nil {
+		return p2p.Headers{
+			"error": []byte("Error creating response pricing headers"),
+		}
+	}
+	s.logger.Debugf("price headler: response target %v with price as %v, for peer %s", chunkAddress, checkPrice, peerAddress)
+
+	return returnHeaders
+
+}
+
+func (s *Pricer) SetKademlia(kad topology.Driver) {
```

Comment:
why is this not set in constructor? is there some circularity?

Reply:
Pricing protocol should exist by the time the first connections are made (it maybe still not be initialized early on enough), otherwise ConnectIn / Out init function does not happen, pricetable/threshold is not announced. Pricing needs pricer, to provide the table, and pricer needs kademlia, to provide neighborhood depth.

So therefore the current order